test/core/common/scale_test.cpp

```diff
@@ -0,0 +1,1226 @@
+          },
+          [](const expected::Error<EncodeError> &) { FAIL(); });
+
+  // encode existing uint16_t
+  advanced::encodeOptional(std::optional<uint16_t>{511})
+      .match(
+          [](const expected::Value<ByteArray> &v) {
+            ASSERT_EQ(v.value, (ByteArray{1, 255, 1}));
+          },
+          [](const expected::Error<EncodeError> &) { FAIL(); });
+
+  // encode existing uint32_t
+  advanced::encodeOptional(std::optional<uint32_t>{67305985})
+      .match(
+          [](const expected::Value<ByteArray> &v) {
+            ASSERT_EQ(v.value, (ByteArray{1, 1, 2, 3, 4}));
+          },
+          [](const expected::Error<EncodeError> &) { FAIL(); });
+}
+
+TEST(Scale, decodeOptional) {
+  // clang-format off
+  auto bytes = ByteArray{
+    0,              // first value
+    1, 1,           // second value
+    1, 255,         // third value
+    0,              // fourth value
+    1, 255, 1,      // fifth value
+    1, 1, 2, 3, 4}; // sixth value
+  // clang-format on
+
+  auto stream = BasicStream{bytes};
+
+  // decode nullopt uint8_t
+  advanced::decodeOptional<uint8_t>(stream).match(
+      [](const expected::Value<std::optional<uint8_t>> &v) {
+        ASSERT_EQ(v.value, std::nullopt);
+      },
+      [](const expected::Error<DecodeError> &) { FAIL(); });
+
+  // decode optional uint8_t
+  advanced::decodeOptional<uint8_t>(stream).match(
+      [](const expected::Value<std::optional<uint8_t>> &v) {
+        ASSERT_EQ(v.value.has_value(), true);
+        ASSERT_EQ((*v.value), 1);
+      },
+      [](const expected::Error<DecodeError> &) { FAIL(); });
+
+  // decode optional negative int8_t
+  advanced::decodeOptional<int8_t>(stream).match(
+      [](const expected::Value<std::optional<int8_t>> &v) {
+        ASSERT_EQ(v.value.has_value(), true);
+        ASSERT_EQ((*v.value), -1);
+      },
+      [](const expected::Error<DecodeError> &) { FAIL(); });
+
+  // decode nullopt uint16_t
+  // it requires 1 zero byte just like any other nullopt
+  advanced::decodeOptional<uint16_t>(stream).match(
+      [](const expected::Value<std::optional<uint16_t>> &v) {
+        ASSERT_EQ(v.value.has_value(), false);
+      },
+      [](const expected::Error<DecodeError> &) { FAIL(); });
+
+  // decode optional uint16_t
+  advanced::decodeOptional<uint16_t>(stream).match(
+      [](const expected::Value<std::optional<uint16_t>> &v) {
+        ASSERT_EQ(v.value.has_value(), true);
+        ASSERT_EQ((*v.value), 511);
+      },
+      [](const expected::Error<DecodeError> &) { FAIL(); });
+
+  // decode optional uint32_t
+  advanced::decodeOptional<uint32_t>(stream).match(
+      [](const expected::Value<std::optional<uint32_t>> &v) {
+        ASSERT_EQ(v.value.has_value(), true);
+        ASSERT_EQ((*v.value), 67305985);
+      },
+      [](const expected::Error<DecodeError> &) { FAIL(); });
+}
+
+// TODO: implement encode and decode tuples and structures
+// TODO: write example of encoding and decoding custom structures
```

Comment:
Is this going to be done in that PR? If not, please create task in write a TODO in a special format

Reply:
This is going to be done in other PR. Task will be created.